src/source/tcp.rs

```diff
@@ -134,44 +139,54 @@ where
             let mut events = mio::Events::with_capacity(1024);
             match poll.poll(&mut events, None) {
                 Err(e) => panic!(format!("Failed during poll {:?}", e)),
-                Ok(_num_events) => {
-                    for event in events {
-                        match event.token() {
-                            constants::SYSTEM => {
-                                for handler in self.handlers {
-                                    handler.shutdown();
-                                }
-
-                                util::send(&mut chans, metric::Event::Shutdown);
-                                return;
+                Ok(_num_events) => for event in events {
+                    match event.token() {
+                        constants::SYSTEM => {
+                            for handler in self.handlers {
+                                handler.shutdown();
                             }
-                            listener_token => {
-                                self.spawn_stream_handlers(
-                                    chans.clone(), // TODO: do not clone, make an Arc
-                                    tags,
-                                    listener_token,
-                                );
+
+                            util::send(&mut chans, metric::Event::Shutdown);
+                            return;
+                        }
+                        listener_token => {
+                            let listener = &(self.listeners[listener_token]);
+                            match self.spawn_stream_handlers(
+                                chans.clone(),
+                                tags,
+                                listener,
+                            ) {
+                                Ok(new_handlers) => {
+                                    self.handlers.extend(new_handlers);
+                                }
+                                Err(e) => {
+                                    error!("Failed to spawn stream handlers! {:?}", e);
+                                    error!("Deregistering listener for {:?} due to unrecoverable error!", *listener);
+                                    let _ = poll.deregister(listener);
+                                }
                             }
                         }
                     }
-                }
+                },
             }
         }
     }
 
     fn spawn_stream_handlers(
-        &mut self,
+        &self,
         chans: util::Channel,
         tags: &sync::Arc<metric::TagMap>,
-        listener_token: mio::Token,
-    ) -> () {
-        let listener = &(self.listeners[listener_token]);
+        listener: &mio::net::TcpListener,
+    ) -> Result<Vec<thread::ThreadHandle>, std::io::Error> {
```

Comment:
This is conveniently aliased `std::io::Result<Vec<thread::ThreadHandler>>`

Reply:
If it is all the same, I prefer as little magic as possible.